pkg/provisioner/ironic/bios.go

```diff
@@ -0,0 +1,171 @@
+	// Build public bios settings
+	settings := buildBIOSSettings(*firmware,
+		map[string]string{
+			"SimultaneousMultithreadingEnabled": "ProcHyperthreading",
+			"VirtualizationEnabled":             "ProcVirtualization",
+			"SriovEnabled":                      "Sriov",
+		},
+		map[string]string{
+			"true":  "Enabled",
+			"false": "Disabled",
+		},
+	)
+	// build private bios settings
+	settings = append(settings, buildBIOSSettings(*firmware.ILO, nil, nil)...)
+
+	// This cleaning step applies a set of BIOS settings for a node
+	cleanSteps = append(
+		cleanSteps,
+		nodes.CleanStep{
+			Interface: "bios",
+			Step:      "apply_configuration",
+			Args: map[string]interface{}{
+				"settings": settings,
+			},
+		},
+	)
+	return
+}
+
+// A private method to build the clean steps for IRMC configuration from BaremetalHost spec
+func buildIRMCCleanSteps(firmware *metal3v1alpha1.FirmwareConfig) (cleanSteps []nodes.CleanStep) {
+	// If not configure irmc, only need to clear old configuration,
+	// but irmc bios interface does not support factory_reset.
+	if firmware == nil || firmware.IRMC == nil {
+		return
+	}
+
+	// Build bios settings
+	// Build public bios settings
+	settings := buildBIOSSettings(*firmware,
+		map[string]string{
+			"SimultaneousMultithreadingEnabled": "hyper_threading_enabled",
+			"VirtualizationEnabled":             "cpu_vt_enabled",
+			"SriovEnabled":                      "single_root_io_virtualization_support_enabled",
+		},
+		nil,
+	)
+	// build private bios settings
+	settings = append(settings, buildBIOSSettings(*firmware.IRMC,
+		map[string]string{
+			"BootOptionFilter":                    "boot_option_filter",
+			"CheckControllersHealthStatusEnabled": "check_controllers_health_status_enabled",
+			"CPUActiveProcessorCores":             "cpu_active_processor_cores",
+			"CPUAdjacentCacheLinePrefetchEnabled": "cpu_adjacent_cache_line_prefetch_enabled",
+			"FlashWriteEnabled":                   "flash_write_enabled",
+			"KeepVoidBootOptionsEnabled":          "keep_void_boot_options_enabled",
+			"LaunchCsmEnabled":                    "launch_csm_enabled",
+			"OsEnergyPerformanceOverrideEnabled":  "os_energy_performance_override_enabled",
+			"PciAspmSupport":                      "pci_aspm_support",
+			"PciAbove4gDecodingEnabled":           "pci_above_4g_decoding_enabled",
+			"PowerOnSource":                       "power_on_source",
+		},
+		nil,
+	)...)
+
+	// This cleaning step applies a set of BIOS settings for a node
+	cleanSteps = append(
+		cleanSteps,
+		nodes.CleanStep{
+			Interface: "bios",
+			Step:      "apply_configuration",
+			Args: map[string]interface{}{
+				"settings": settings,
+			},
+		},
+	)
+	return
+}
+
+// A private method to build bios settings from different driver
+func buildBIOSSettings(biosSettings interface{}, nameMap map[string]string, valueMap map[string]string) (settings []map[string]string) {
+	var value string
+	var name string
+	v := reflect.ValueOf(biosSettings)
```

Comment:
Could we design the class structure to avoid reflection? E.g. make each BMC class convert the generic structure to string values?

Reply:
Could you give me an example? In my understanding, I may need to add a lot of 'if' to avoid using reflection.

Reply:
> Could we design the class structure to avoid reflection? E.g. make each BMC class convert the generic structure to string values?

Wouldn't we need generic names for the fields in that case? (for the generic structure) An agreed upon naming convention?

Also, what are the drawbacks of using reflection?